Custom Error Message Support and improve test coverage

ganeshpanruti opened the pull request.

neoremind commented on fluent-validator/src/main/java/com/baidu/unbiz/fluentvalidator/FluentValidator.java:
```diff
@@ -527,6 +540,67 @@ public FluentValidator doValidate(ValidateCallback cb) {
     }
 
+    
+    public FluentValidator doCustomValidate(ValidateCallback cb) {
```
Copying the whole method but only left on parameter difference, this is not acceptable because it breaks common DRY(do not repeat yourself) rule. So could you refine it a better a way? maybe we can leverage varargs?

Also one more question, what is your purpose of specifying a error message to "on" method? I do not see the usage in any test cases. Could you provide more information? And maybe we can come up a better solution.

For you PR, I suggest you use it on your local maven repo first. Once we settle down, I will publish it to central repo asap.

ganeshpanruti commented:
Hi Neoremind,

I have added new TestCases for this and removed the code repeated.
The main uses is , if im am having a Validator called CustomNullCheckValidator, in on method i will pass the custom message so if it is null it will throw the custom message.
If you see the test cases i can get clear about that.

  Car c = new Car(null,null, 10);
       
        Result ret = FluentValidator.checkAll().failOver()
                .on(c,new CustomNullCheckValidator(),"Car Object is Null")
                .on(c.getLicensePlate() ,new CustomNullCheckValidator(),"Car License is Null")
                .on(c.getManufacturer(),new CustomNullCheckValidator(),"Car Manufactur is Null")
                .on(c.getSeatCount())
                .doValidate(true)
                .result(toSimple());
        System.out.println(ret);
        assertThat(ret.isSuccess(), is(false));
        assertEquals(ret.errors.toString(),"[Car License is Null, Car Manufactur is Null]");

neoremind commented:
That looks great! I got your purpose for adding the feature. But I notice there are some code style issue that needs to be refined, like there should be one space after comma, brace should be in the same line with if clause. I will review the code first and accept PR, after that I will do some slightly changes if possible to make in concise, then publish to maven central. 

BTW, are you in hurry to have it delivered?

ganeshpanruti commented:
Ya neoremind. It would be good if you do it as soon as possible. Thanks a lot for adding the changes.

ganeshpanruti commented:
I'm in hurry to get the changes. It would be good if i get it early as possible. Thanks.

ganeshpanruti commented:
Hi @neoremind ,

Any update on the merge, When can i expect this to be merged.

Regards,
Ganesh

neoremind commented:
Hi ganeshpanruti,

Sorry about the delay, I have been super busy lately, and after re-thinking I think this change would affect the core Validator interface definition which would probably affect existing end-users' code. I am thinking about a less api-change-prone solution and will come up idea later and discuss with you if it will meet your requirement. The time would be at the end of next wee. Is it work for you ?

Thanks,
Xu

ganeshpanruti commented:
Sure neoremind, We can have a discussion on that...

neoremind commented:
got it, will review soon. thx

neoremind commented:
Hi ganeshpanruti,
 
I have finished your requirement about customizing error message. I created a PR myself, you can find it from the link [https://github.com/neoremind/fluent-validator/pull/17](https://github.com/neoremind/fluent-validator/pull/17).

Here I will explain my implementation, after reviewing your PR, I found you made changes to the `Validator` interface by specifying another `validate` method with error message as one of the params, my concern is that other users who does not rely on `ValidatorHandler` (which is a subclass of `Validator`) will suffer from API changes if they upgrade their library, so they have to implement another `validate(target, message)` method in a forced way.

In order to meet your requirement and avoid break the definition of the interface. I initiated a workaround to inject error message from where `FluentValidator` is invoked, and also `Validator` can use the error message in ease. For example,
```
        Car c = new Car(null, null, 10);
        Result ret = FluentValidator.checkAll().failOver()
                .on(c, new CustomNullCheckValidator(), "Car Object is Null")
                .on(c.getLicensePlate(), new CustomNullCheckValidator(), "Car License is Null")
                .on(c.getManufacturer(), new CustomNullCheckValidator(), "Car Manufacture is Null")
                .on(c.getSeatCount())
                .doValidate()
                .result(toSimple());
        System.out.println(ret);
        assertThat(ret.isSuccess(), is(false));
        assertThat(ret.errors, is((List) Lists.newArrayList(
                "Car License is Null",
                "Car Manufacture is Null"
        )));
```

your validator looks like:
```
public class CustomNullCheckValidator extends ValidatorHandler<Object> implements Validator<Object> {

    @Override
    public boolean validate(ValidatorContext context, Object t) {
        if (t == null) {
            context.addErrorMsg(context.getDefaultMessage());
            return false;
        }
        return true;
    }
}
```

you can get message from context by invoking `getDefaultMessage()` method.

Please let me know if this change works for you. If you think it is OK, then I will release 1.0.7 version to maven central so that you guys can use it.

Regards,
Xu

ganeshpanruti commented:
Hi Xu,

It looks absolute great. When can i expect that release version. 

Thanks,
Ganesh.V

neoremind commented:
Hi ganeshpanruti,

Great! I expect to release it to maven central tomorrow. I will let you know as soon as it is done.

Thanks,
Xu
